### PR TITLE
VIM-6867: Fix build errors in Swift 4

### PR DESCRIPTION
**Pull request by Jumhyn**
#### Ticket

[VIM-6867](https://vimean.atlassian.net/browse/VIM-6867)

#### Pull Request Checklist

- [ ] Resolved any merge conflicts
- [ ] No build errors or warnings are introduced
- [ ] New files are written in Swift
- [ ] New classes contain license headers
- [ ] New classes have Documentation
- [ ] New public methods have Documentation

#### Issue Summary

- VimeoUpload cannot build with the Swift 4 compiler.

#### Implementation Summary

- Add `@objc` to methods used as arguments to `#selector`.

- Add `@objc` to all `dynamic` vars.

- Add new convenience `init` to `VideoDeletionManager` to move private constant out of default argument position.

- Change cast type in `NSError+Extensions`.

- In `UploadDescriptor`, replace references to `type(of:)` with `UploadDescriptor` to avoid ambiguity with `type` variable requirement from `Descriptor` protocol. Since we are only referencing `private static` constants this should have no impact on inheritance behavior now or down the line.

- Update for API changes in `ExportOperation`.

#### How to Test

### CODE REVIEW

**Comment by Jumhyn** [comment]
@jasonhawkins Not sure what the implications are around this, but had some extra time today so figured it would be useful. We'd have to do a major release presumably, right? Will clients have to update the Swift version manually, or will they pull it from the project settings here?

**Comment by jasonhawkins** [comment]
This is awesome and desperately needed, so thank you for jumping on it. But yeah, this would definitely warrant a major release. Clients that were mixing Swift 3.2 and 4.1 modules would have to include some post-install instructions in their Podfile to declare that a specific target should get Swift 4.1. But yeah, that'll be up to the client. 

A couple other things:
- It looks like we'll need to update all of the projects in the workspace, **VimeoUpload**, **VimeoUpload-iOS-OldUpload**, and **VimeoUpload-iOS** to use Swift 4.1. I was getting an error on `bundle exec pod install`.
- There are likely a bunch of other interfaces we'll need to explicitly add the `@objc` attribute to. I tried building the `VimoeUpload-iOS` target and ran into a handful of build warnings for that reason. This might be huge pain 😬
- It looks like we're pointing to VimeoNetworking as a submodule here, and not as a pod. I feel like we should probably go ahead and add it in our Podfile and remove it as a submodule. But maybe that's a separate ticket?
- We should definitely consider / discuss pushing this to CocoaPods when we merge this work in.

This is great, and I'm excited we're doing this 🎉

File: VimeoUpload/UIKit/DescriptorKVObserver.swift [hunk 1]
```diff
@@ -64,7 +64,7 @@ public class DescriptorKVObserver: NSObject
         {
             if let descriptor = self.descriptor
             {
-                let type = descriptor.type
+                let type = descriptor.descriptorType
```
Comment (nicolelehrer) [inline VimeoUpload/UIKit/DescriptorKVObserver.swift]: Need to make sure this isn't a breaking change. Had to rename `type` -> `descriptorType` due to compiler confusion with global namespace function `type(of:)`

**Comment by nicolelehrer** [comment]
Other remaining **TODO**: test for Swift 3 @objc inference using “Use of deprecated Swift 3 @objc inference” logging enabled, and then disable inference by changing the "Swift 3 @objc Inference" build setting to "Default" for the relevant targets

File: VimeoUpload/UIKit/DescriptorKVObserver.swift (hunk 1 above)
Comment (nicolelehrer) [reply VimeoUpload/UIKit/DescriptorKVObserver.swift]: This caused no issues uploading using old or new upload. Will be tested in context of client in https://vimean.atlassian.net/browse/VIM-6844

**Comment by nicolelehrer** [comment]
Thank you @jasonhawkins 
- I added your suggestions for the podfile and the consumption of `@objc` API by our client 
- Also went tagged with `@objc` any `public` or `open` API that could be visible to Objective-C like we discussed
- Ticketed the update to podspec to handle macOS so we give it proper time to test  before we formally support https://vimean.atlassian.net/browse/VIM-6877
- Ticketed releases for all libraries with 3.x https://vimean.atlassian.net/browse/VIM-6876 before we merge feature branch into develop

**Pull request merged**